daemon/src/main/java/org/mvndaemon/mvnd/daemon/Server.java

```diff
@@ -246,6 +260,23 @@ private void client(SocketChannel socket) {
         }
     }
 
+    private boolean checkToken(SocketChannel socket) {
+        byte[] token = new byte[info.getToken().length];
+        ByteBuffer tokenBuffer = ByteBuffer.wrap(token);
+
+        try {
+            do {
+                if (socket.read(tokenBuffer) == -1) {
+                    break;
+                }
+            } while (tokenBuffer.remaining() > 0);
+        } catch (final IOException e) {
+            LOGGER.debug("Discarding EOFException: {}", e.toString(), e);
+        }
+
+        return Arrays.equals(info.getToken(), token);
```

Comment:
This is vulnerable to timing attacks, since the comparison will stop at the first non-equal byte. An attacker can use this to learn the token. Instead, use [MessageDigest.isEqual()](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/security/MessageDigest.html#isEqual(byte%5B%5D,byte%5B%5D)) which does a constant time comparison (does not depend on the array contents).

Reply:
Thanks for the hint!
I fixed it.